File: src/Pybind11Wraps/CXXTypes/SpheralCXXTypes.cc
// Put Python includes first to avoid compile warnings about redefining _POSIX_C_SOURCE
#include "pybind11/pybind11.h"
#include "pybind11/stl_bind.h"
#include "pybind11/operators.h"

#include <vector>
#include <map>
#include <set>
#include <string>

#include <limits.h>

namespace py = pybind11;

// vector
PYBIND11_MAKE_OPAQUE(std::vector<char>);
PYBIND11_MAKE_OPAQUE(std::vector<unsigned>);
PYBIND11_MAKE_OPAQUE(std::vector<uint64_t>);
PYBIND11_MAKE_OPAQUE(std::vector<int>);
PYBIND11_MAKE_OPAQUE(std::vector<float>);
PYBIND11_MAKE_OPAQUE(std::vector<double>);
PYBIND11_MAKE_OPAQUE(std::vector<std::string>);

// vector<vector<>>
PYBIND11_MAKE_OPAQUE(std::vector<std::vector<char>>);
PYBIND11_MAKE_OPAQUE(std::vector<std::vector<unsigned>>);
PYBIND11_MAKE_OPAQUE(std::vector<std::vector<uint64_t>>);
PYBIND11_MAKE_OPAQUE(std::vector<std::vector<int>>);
PYBIND11_MAKE_OPAQUE(std::vector<std::vector<float>>);
PYBIND11_MAKE_OPAQUE(std::vector<std::vector<double>>);
PYBIND11_MAKE_OPAQUE(std::vector<std::vector<std::string>>);

typedef std::pair<double, double> pair_double_double;
typedef std::pair<double, std::string> pair_double_string;
typedef std::pair<unsigned, unsigned> pair_unsigned_unsigned;
typedef std::pair<uint64_t, uint64_t> pair_ULL_ULL;
typedef std::pair<std::string, std::string> pair_string_string;
PYBIND11_MAKE_OPAQUE(std::vector<pair_double_double>);
PYBIND11_MAKE_OPAQUE(std::vector<pair_double_string>);
PYBIND11_MAKE_OPAQUE(std::vector<pair_unsigned_unsigned>);
PYBIND11_MAKE_OPAQUE(std::vector<pair_ULL_ULL>);
PYBIND11_MAKE_OPAQUE(std::vector<pair_string_string>);

// map
typedef std::map<std::string, double> map_string_double;
typedef std::map<int, std::string> map_int_string;
PYBIND11_MAKE_OPAQUE(map_string_double);
PYBIND11_MAKE_OPAQUE(map_int_string);

// Make the module
PYBIND11_PLUGIN(SpheralCXXTypes) {
  py::module m("SpheralCXXTypes", "Spheral C++ types module.");

  // vector
  py::bind_vector<std::vector<char>>(m, "vector_of_char");
  py::bind_vector<std::vector<unsigned int>>(m, "vector_of_unsigned");
  py::bind_vector<std::vector<uint64_t>>(m, "vector_of_ULL");
  py::bind_vector<std::vector<int>>(m, "vector_of_int");
  py::bind_vector<std::vector<float>>(m, "vector_of_float");
  py::bind_vector<std::vector<double>>(m, "vector_of_double");
  py::bind_vector<std::vector<std::string>>(m, "vector_of_string");

  // vector<vector<>>
  py::bind_vector<std::vector<std::vector<char>>>(m, "vector_of_vector_of_char");
  py::bind_vector<std::vector<std::vector<unsigned int>>>(m, "vector_of_vector_of_unsigned");
  py::bind_vector<std::vector<std::vector<uint64_t>>>(m, "vector_of_vector_of_ULL");
  py::bind_vector<std::vector<std::vector<int>>>(m, "vector_of_vector_of_int");
  py::bind_vector<std::vector<std::vector<float>>>(m, "vector_of_vector_of_float");
  py::bind_vector<std::vector<std::vector<double>>>(m, "vector_of_vector_of_double");
  py::bind_vector<std::vector<std::vector<std::string>>>(m, "vector_of_vector_of_string");

  // vector<pair>
  py::bind_vector<std::vector<std::pair<double, double>>>(m, "vector_of_pair_double_double");
  py::bind_vector<std::vector<std::pair<double, std::string>>>(m, "vector_of_pair_double_string");
  py::bind_vector<std::vector<std::pair<unsigned, unsigned>>>(m, "vector_of_pair_unsigned_unsigned");
  py::bind_vector<std::vector<std::pair<uint64_t, uint64_t>>>(m, "vector_of_pair_ULL_ULL");
  py::bind_vector<std::vector<std::pair<std::string, std::string>>>(m, "vector_of_pair_string_string");

  // map
  py::bind_map<std::map<std::string, double>>(m, "map_string_double");
  py::bind_map<std::map<int, std::string>>(m, "map_int_string");

  return m.ptr();
}

File: src/Pybind11Wraps/CXXTypes/SpheralCXXTypes.py
"""
Spheral CXXTypes module.

Provides access to fundamental C++ types from python that are not Spheral specific.
"""

from PYB11Generator import *

# preamble = """
# typedef std::pair<double, double> pair_double_double;
# typedef std::pair<double, std::string> pair_double_string;
# typedef std::pair<unsigned, unsigned> pair_unsigned_unsigned;
# typedef std::pair<uint64_t, uint64_t> pair_ULL_ULL;
# typedef std::pair<std::string, std::string> pair_string_string;
# """

# Include files.
includes = ["<vector>",
            "<map>",
            "<set>",
            "<string>"]

# std::vector
vector_of_char     = PYB11_bind_vector("char", opaque=True)
vector_of_unsigned = PYB11_bind_vector("unsigned", opaque=True)
vector_of_ULL      = PYB11_bind_vector("uint64_t", opaque=True)
vector_of_int      = PYB11_bind_vector("int", opaque=True)
vector_of_float    = PYB11_bind_vector("float", opaque=True)
vector_of_double   = PYB11_bind_vector("double", opaque=True)
vector_of_string   = PYB11_bind_vector("std::string", opaque=True)

# std::vector<std::vector>
vector_of_vector_of_char     = PYB11_bind_vector("std::vector<char>", opaque=True)
vector_of_vector_of_unsigned = PYB11_bind_vector("std::vector<unsigned>", opaque=True)
vector_of_vector_of_ULL      = PYB11_bind_vector("std::vector<uint64_t>", opaque=True)
vector_of_vector_of_int      = PYB11_bind_vector("std::vector<int>", opaque=True)
vector_of_vector_of_float    = PYB11_bind_vector("std::vector<float>", opaque=True)
vector_of_vector_of_double   = PYB11_bind_vector("std::vector<double>", opaque=True)
vector_of_vector_of_string   = PYB11_bind_vector("std::vector<std::string>", opaque=True)

# std::vector<pair<>>
vector_of_pair_double_double     = PYB11_bind_vector("pair<double,_double>", opaque=True)
# vector_of_pair_double_string     = PYB11_bind_vector("pair_double_string", opaque=True)
# vector_of_pair_unsigned_unsigned = PYB11_bind_vector("pair_unsigned_unsigned", opaque=True)
# vector_of_pair_ULL_ULL           = PYB11_bind_vector("pair_ULL_ULL", opaque=True)
# vector_of_pair_string_string     = PYB11_bind_vector("pair_string_string", opaque=True)

# std::map
map_string_double = PYB11_bind_map("std::string", "double", opaque=True)
map_int_string    = PYB11_bind_map("int", "std::string", opaque=True)
